src/hotspot/share/prims/jvmtiRedefineClasses.cpp

```diff
@@ -3543,32 +3543,48 @@ void VM_RedefineClasses::set_new_constant_pool(
   int i;  // for portability
 
   // update each field in klass to use new constant pool indices as needed
-  for (JavaFieldStream fs(scratch_class); !fs.done(); fs.next()) {
-    jshort cur_index = fs.name_index();
+  int java_fields;
+  int injected_fields;
+  bool update_required = false;
+  GrowableArray<FieldInfo>* fields = FieldInfoStream::create_FieldInfoArray(scratch_class->fieldinfo_stream(), &java_fields, &injected_fields);
+  for (int i = 0; i < java_fields; i++) {
+    FieldInfo* fi = fields->adr_at(i);
+    jshort cur_index = fi->name_index();
     jshort new_index = find_new_index(cur_index);
     if (new_index != 0) {
       log_trace(redefine, class, constantpool)("field-name_index change: %d to %d", cur_index, new_index);
-      fs.set_name_index(new_index);
+      fi->set_name_index(new_index);
+      update_required = true;
     }
-    cur_index = fs.signature_index();
+    cur_index = fi->signature_index();
     new_index = find_new_index(cur_index);
     if (new_index != 0) {
       log_trace(redefine, class, constantpool)("field-signature_index change: %d to %d", cur_index, new_index);
-      fs.set_signature_index(new_index);
+      fi->set_signature_index(new_index);
+      update_required = true;
     }
-    cur_index = fs.initval_index();
+    cur_index = fi->initializer_index();
     new_index = find_new_index(cur_index);
     if (new_index != 0) {
       log_trace(redefine, class, constantpool)("field-initval_index change: %d to %d", cur_index, new_index);
-      fs.set_initval_index(new_index);
+      fi->set_initializer_index(new_index);
+      update_required = true;
     }
-    cur_index = fs.generic_signature_index();
+    cur_index = fi->generic_signature_index();
     new_index = find_new_index(cur_index);
     if (new_index != 0) {
       log_trace(redefine, class, constantpool)("field-generic_signature change: %d to %d", cur_index, new_index);
-      fs.set_generic_signature_index(new_index);
+      fi->set_generic_signature_index(new_index);
+      update_required = true;
     }
-  } // end for each field
+  }
+  if (update_required) {
+    Array<u1>* old_stream = InstanceKlass::cast(scratch_class)->fieldinfo_stream();
```

Comment:
scratch_class should already be an InstanceKlass, ie cast not required here or below.